File: backend/app/services/nf_valores.py
"""Cálculo fiscal de contas a receber (imposto e líquido por alíquota)."""
from fastapi import HTTPException


def calcular_imposto_liquido(
    valor_bruto: float,
    aliquota_pct: float | None,
) -> tuple[float, float, float | None]:
    """Retorna (valor_imposto, valor_liquido, aliquota_gravada)."""
    if aliquota_pct is None:
        pct = 0.0
        aliq_gravada = None
    else:
        pct = float(aliquota_pct)
        aliq_gravada = pct
    if pct < 0 or pct > 100:
        raise HTTPException(status_code=400, detail="Alíquota inválida (deve estar entre 0 e 100)")
    imposto = round(valor_bruto * (pct / 100), 2)
    liquido = round(valor_bruto - imposto, 2)
    return imposto, liquido, aliq_gravada

File: frontend/src/utils/nfValores.ts
import type { ContaCorrente } from '../types';
import { codigoPadrao } from './fluxoCaixaMovimentos';

export function calcularImpostoLiquido(
  bruto: number,
  aliquotaPct: number | null | undefined,
): { imposto: number; liquido: number } {
  const pct = aliquotaPct == null || Number.isNaN(aliquotaPct) ? 0 : aliquotaPct;
  if (Number.isNaN(bruto) || bruto < 0) {
    return { imposto: 0, liquido: 0 };
  }
  const imposto = Math.round(bruto * (pct / 100) * 100) / 100;
  const liquido = Math.round((bruto - imposto) * 100) / 100;
  return { imposto, liquido };
}

export function codigoSlot1(contas: ContaCorrente[]): string {
  const ativas = contas.filter((c) => c.ativo);
  return ativas[0]?.codigo || codigoPadrao(contas);
}

export function validarAliquota(valor: string): string | null {
  if (valor.trim() === '') return null;
  const n = parseFloat(valor);
  if (Number.isNaN(n) || n < 0 || n > 100) {
    return 'Alíquota deve estar entre 0 e 100';
  }
  return null;
}

export function aplicarCalculoFiscal(brutoStr: string, aliqStr: string): {
  valor_imposto: string;
  valor_liquido: string;
} {
  const bruto = parseFloat(brutoStr);
  if (!brutoStr.trim() || Number.isNaN(bruto)) {
    return { valor_imposto: '', valor_liquido: '' };
  }
  const aliq = aliqStr.trim() === '' ? null : parseFloat(aliqStr);
  const { imposto, liquido } = calcularImpostoLiquido(bruto, aliq);
  return { valor_imposto: String(imposto), valor_liquido: String(liquido) };
}
